model/util.go

```diff
@@ -17,114 +17,256 @@ package model
 )
 
 // +k8s:deepcopy-gen=false
 
-const prefix = "file:/"
+type Kind interface {
+	KindValues() []string
+	String() string
+}
 
-// TRUE used by bool fields that needs a boolean pointer
-var TRUE = true
+// TODO: Remove global variable
+var httpClient = http.Client{Timeout: time.Duration(1) * time.Second}
 
-// FALSE used by bool fields that needs a boolean pointer
-var FALSE = false
+type UnmarshalError struct {
+	err           error
+	parameterName string
+	primitiveType reflect.Kind
+	objectType    reflect.Kind
+}
 
-func getBytesFromFile(s string) (b []byte, err error) {
-	// #nosec
-	if resp, err := http.Get(s); err == nil {
-		defer resp.Body.Close()
-		buf := new(bytes.Buffer)
-		if _, err = buf.ReadFrom(resp.Body); err != nil {
-			return nil, err
+func (e *UnmarshalError) Error() string {
+	if e.err == nil {
+		panic("unmarshalError fail")
+	}
+
+	var syntaxErr *json.SyntaxError
+	var unmarshalTypeErr *json.UnmarshalTypeError
+	if errors.As(e.err, &syntaxErr) {
+		return fmt.Sprintf("%s has a syntax error %q", e.parameterName, syntaxErr.Error())
+
+	} else if errors.As(e.err, &unmarshalTypeErr) {
+		return e.unmarshalMessageError(unmarshalTypeErr)
+	}
+
+	return e.err.Error()
+}
+
+func (e *UnmarshalError) unmarshalMessageError(err *json.UnmarshalTypeError) string {
+	if err.Struct == "" && err.Field == "" {
+		primitiveTypeName := e.primitiveType.String()
+		var objectTypeName string
+		if e.objectType != reflect.Invalid {
+			switch e.objectType {
+			case reflect.Struct:
+				objectTypeName = "object"
+			case reflect.Map:
+				objectTypeName = "object"
+			case reflect.Slice:
+				objectTypeName = "array"
+			default:
+				objectTypeName = e.objectType.String()
+			}
 		}
-		return buf.Bytes(), nil
-	}
-	s = strings.TrimPrefix(s, prefix)
-
-	if !filepath.IsAbs(s) {
-		// The import file is an non-absolute path, we join it with include path
-		// TODO: if the file didn't find in any include path, we should report an error
-		for _, p := range IncludePaths() {
-			sn := filepath.Join(p, s)
-			if _, err := os.Stat(sn); err == nil {
-				s = sn
-				break
+		return fmt.Sprintf("%s must be %s or %s", e.parameterName, primitiveTypeName, objectTypeName)
+
+	} else if err.Struct != "" && err.Field != "" {
+		var primitiveTypeName string
+		val := reflect.New(err.Type)
+		if valKinds, ok := val.Elem().Interface().(validator.Kind); ok {
+			values := valKinds.KindValues()
+			if len(values) <= 2 {
+				primitiveTypeName = strings.Join(values, " or ")
+			} else {
+				primitiveTypeName = fmt.Sprintf("%s, %s", strings.Join(values[:len(values)-2], ", "), strings.Join(values[len(values)-2:], " or "))
 			}
+		} else {
+			primitiveTypeName = err.Type.Name()
 		}
+
+		return fmt.Sprintf("%s.%s must be %s", e.parameterName, err.Field, primitiveTypeName)
 	}
 
-	if b, err = os.ReadFile(filepath.Clean(s)); err != nil {
+	return err.Error()
+}
+
+func getBytesFromFile(uri string) (b []byte, err error) {
+	refUrl, err := url.Parse(uri)
```

Comment:
If no schema, then `file`? And then we can parse. So we won't require users to add the schema to their paths.